### PR TITLE
add unit tests for the cape feature extractor

### PR DESCRIPTION
**Pull request by yelhamer**
addresses #1498 and #1546 

### Checklist

- [ ] No CHANGELOG update needed

- [x] No new tests needed

- [x] No documentation update needed

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -5,6 +5,7 @@
 
 ### New Features
 - Utility script to detect feature overlap between new and existing CAPA rules [#1451](https://github.com/mandiant/capa/issues/1451) [@Aayush-Goel-04](https://github.com/aayush-goel-04)
+- Add unit tests for the new CAPE extractor @yelhamer
```
Comment (mr-tz) [inline CHANGELOG.md:8]:
```suggestion
- Add unit tests for the new CAPE extractor #1563 @yelhamer
```

Comment (mr-tz) [reply CHANGELOG.md:8]: adding issue reference

File: tests/fixtures.py [hunk 2]
```diff
@@ -183,6 +183,20 @@ def get_binja_extractor(path):
...
 
+@lru_cache(maxsize=1)
+def get_cape_extractor(path):
+    import json
+
+    from capa.features.extractors.cape.extractor import CapeExtractor
+
+    with open(path) as report_file:
+        report = report_file.read()
+        report = json.loads(report)
```
Comment (mr-tz) [inline tests/fixtures.py:194]: consider adding explicit arguments ("r", encoding="utf-8")

File: tests/fixtures.py [hunk 3]
```diff
@@ -183,6 +183,20 @@ def get_binja_extractor(path):
...
+
+    extractor = CapeExtractor.from_report(report)
+    return extractor
```
Comment (mr-tz) [inline tests/fixtures.py:197]:
```suggestion
    return CapeExtractor.from_report(report)
```

File: tests/fixtures.py [hunk 4]
```diff
@@ -491,6 +540,40 @@ def inner_function(extractor):
 
         inner_function.__name__ = scope
         return inner_function
+    elif "thread=" in scope:
+        # like `process=(pid:ppid),thread=1002`
+        assert "process=" in scope
+        pspec, _, tspec = scope.partition(",")
+        pspec = pspec.partition("=")[2][1:-1].split(":")
+        assert len(pspec) == 2
+        pid, ppid = map(lambda x: int(x), pspec)
```
Comment (mr-tz) [inline tests/fixtures.py:549]:
```suggestion
        pid, ppid = map(int, pspec)
```

File: tests/fixtures.py [hunk 5]
```diff
@@ -491,6 +540,40 @@ def inner_function(extractor):
...
+                features[k].update(vs)
+            return features
+
+        inner_thread.__name__ = scope
+        return inner_thread
+    elif "process=" in scope:
+        # like `process=(pid:ppid)`
+        pspec = scope.partition("=")[2][1:-1].split(":")
+        assert len(pspec) == 2
+        pid, ppid = map(lambda x: int(x), pspec)
```
Comment (mr-tz) [inline tests/fixtures.py:566]:
```suggestion
        pid, ppid = map(int, pspec)
```

**Review by mr-tz — COMMENTED** [review]
nice progress, a few minor comments

**Comment by yelhamer** [comment]
The requested changes should be addressed now. I am unsure however why code_style tests are failing for base_extractor.py here but not in the other PR.

File: tests/fixtures.py [hunk 6]
```diff
@@ -311,6 +342,10 @@ def get_data_path_by_name(name):
         return os.path.join(CD, "data", "294b8db1f2702b60fb2e42fdc50c2cee6a5046112da9a5703a548a4fa50477bc.elf_")
     elif name.startswith("2bf18d"):
         return os.path.join(CD, "data", "2bf18d0403677378adad9001b1243211.elf_")
+    elif name.startswith("0000a657"):
+        return os.path.join(
+            CD, "data/dynamic/cape", "0000a65749f5902c4d82ffa701198038f0b4870b00a27cfca109f8f933476d82.json.gz"
```
Comment (williballenthin) [inline tests/fixtures.py:347]:
```suggestion
            CD, "data", "dynamic", "cape", "0000a65749f5902c4d82ffa701198038f0b4870b00a27cfca109f8f933476d82.json.gz"
```

we try not to refer to the path separator directly, so things work well on linux/windows

**Review by williballenthin — APPROVED** [review]
from reading the code, this PR appears good. but i'd like to see it working against the test data.

would you merge this branch with your existing `cape-extractor` branch and PR and then we'll see how all the code works together? in the future, tests should be provided alongside the feature PR so that CI/CD can show that the new code is working.

File: tests/fixtures.py (hunk 6 above)
Comment (mr-tz) [inline tests/fixtures.py:347]: let's account for having the same file in here for static and dynamic, I think as of now this could break here

**Comment by williballenthin** [comment]
lets close this PR because the tests are added in #1546

**Pull request closed without merging**